### PR TITLE
Fixed total number of batches

### LINKED ISSUE #430: Incorrect total number of batches
**Describe the bug**
Sometimes total number of batches is computed wrong.

**To Reproduce**
Run the following code with current master branch:
```
from time import sleep
import torch
from torch.utils.data import DataLoader, Dataset

import pytorch_lightning as pl


class DummyDataset(Dataset):
    def __init__(self, n):
        super().__init__()
        self.n = n

    def __len__(self):
        return self.n

    def __getitem__(self, idx):
        return torch.rand(10)


class CoolSystem(pl.LightningModule):
    def __init__(self):
        super(CoolSystem, self).__init__()
        self.layer = torch.nn.Linear(10, 10)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_nb):
        sleep(1)
        return {'loss': torch.mean(self.forward(batch) ** 2)}

    def validation_step(self, batch, batch_nb):
        sleep(1)
        return {}

    def validation_end(self, outputs):
        return {}

    def configure_optimizers(self):
        return [torch.optim.Adam(self.layer.parameters())]

    @pl.data_loader
    def train_dataloader(self):
        return DataLoader(DummyDataset(10), batch_size=1)

    @pl.data_loader
    def val_dataloader(self):
        return DataLoader(DummyDataset(5), batch_size=1)

model = CoolSystem()
trainer = pl.Trainer(weights_summary=None, nb_sanity_val_steps=0, early_stop_callback=False,
                     val_percent_check=1.0, val_check_interval=0.5)
trainer.fit(model)
```
At first output will look like:
`67%|█████▋                    | 10/15 [00:10<00:05,  1.05s/it, batch_nb=4, epoch=0, loss=0.194, v_nb=0]`

But at the end of the epoch it will be like:
`20it [00:20,  1.07s/it, batch_nb=9, epoch=0, loss=0.212, v_nb=0]`

Moreover, if you run
```
trainer = pl.Trainer(weights_summary=None, nb_sanity_val_steps=0, early_stop_callback=False,
                     val_percent_check=1.0, val_check_interval=0.5, check_val_every_n_epoch=10)
```
The first epoch will end at the point 
`67%|█████▋                    | 10/15 [00:09<00:04,  1.01it/s, batch_nb=8, epoch=1, loss=0.069, v_nb=0]` 

**Expected behavior**
Correct total number of batches.

**Possible solution**
Now we have `total_batches = nb_training_batches + nb_val_batches`, where `nb_val_batches` is the number of batches of only one validation loop. And the problem arises because actually there can be several validation loops during one training epoch. Moreover there is a parameter `check_val_every_n_epoch` and thus there can be no validation loops at all.

With this in mind, it looks like the correct formula is:
```
is_val_epoch = (current_epoch + 1) % check_val_every_n_epoch == 0
val_checks_per_epoch = nb_training_batches // val_check_batch if is_val_epoch else 0
total_batches = nb_training_batches + nb_val_batches * val_checks_per_epoch
```

### PR DESCRIPTION
**Pull request by kuynzereb**
Fixes #430

### CODE REVIEW

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 1]
```diff
@@ -23,7 +23,15 @@ def train(self):
             # update training progress in trainer and model
             model.current_epoch = epoch_nb
             self.current_epoch = epoch_nb
-            self.total_batches = self.nb_training_batches + self.nb_val_batches
+
+            # val can be checked multiple times in epoch
+            is_val_epoch = (self.current_epoch + 1) % self.check_val_every_n_epoch == 0
+            val_checks_per_epoch = self.nb_training_batches // self.val_check_batch
```
Comment (Borda) [inline pytorch_lightning/trainer/train_loop_mixin.py]: `self.val_check_batch` is a number of batches used for validation?

Comment (kuynzereb) [reply pytorch_lightning/trainer/train_loop_mixin.py]: Nope. It is the number of a training batch when validation loop should be run, i. e., we run validation loop in the end of the current training iteration if `(batch_nb + 1) % self.val_check_batch == 0`.

Comment (Borda) [reply pytorch_lightning/trainer/train_loop_mixin.py]: I understand the `is_val_epoch` but still a bit hard to follow the remaining, why you divide  `nb_training_batches` (number of training batches in epoch) by `val_check_batch` (number of validation batches)...

Comment (kuynzereb) [reply pytorch_lightning/trainer/train_loop_mixin.py]: `val_check_batch` is not the number of validation batches. We run validation loop, when current training batch is equal to `val_check_batch`, `2 * val_check_batch`, `3 * val_check_batch` and so on. Thus during one training epoch we will run validation loop `nb_training_batches // val_check_batch` times. For example, if `val_check_batch = nb_training_batches` then we will run validation loop only once per epoch, at the very end of this epoch. And if `val_check_batch = 1` then we will run validation loop after every training batch, so there will be `nb_training_batches`

Comment (Borda) [reply pytorch_lightning/trainer/train_loop_mixin.py]: so `val_check_batch` is the number of validation batches per epoch which are equally distributed

Comment (kuynzereb) [reply pytorch_lightning/trainer/train_loop_mixin.py]: I look at this from the other angle. By setting `nb_checks` you control how often you want to check your validation metrics. By setting `val_percent_check` you control how accurate your validation estimates. Ideally, you should set `val_percent_check=1` to obtain most accurate validation estimates. But in fact validation set may be quite big, so it would be very time consuming to use all validation data for one validation estimate. So you reduce `val_percent_check` to obtain less accurate but faster estimates.

So for me it is strange to distribute validation batches because total number of validation batches per epoch has no real meaning. Whereas `nb_checks` controls how often you make validation checks, and `val_percent_check` controls how accurate/fast your estimates.

Comment (Borda) [reply pytorch_lightning/trainer/train_loop_mixin.py]: ok, correct me if I am wrong, just imagine that you have available 10 batches for validation but you use just 2 of them to be faster, and then you in some moment say I want to repeat it twice in each epoch... then you will repeat just the 2 (of 10) batches or will you take in fact another (4 of 10)?

Comment (kuynzereb) [reply pytorch_lightning/trainer/train_loop_mixin.py]: So answering your question:
> Just curious what is the reason running validation check of full size (eg 5 batches) multiple times in single epoch

Here is two things. You may want to run validation several times during one epoch if you have a lot of training data, so your epochs are very big but you want to check validation metrics more often. And you may want to compute these metrics using all validation data to obtain most accurate results.

Comment (kuynzereb) [reply pytorch_lightning/trainer/train_loop_mixin.py]: > ok, correct me if I am wrong, just imagine that you have available 10 batches for validation but you use just 2 of them to be faster, and then you in some moment say I want to repeat it twice in each epoch... then you will repeat just the 2 (of 10) batches or will you take in fact another (4 of 10)?

I guess it depends on how you have implemented `val_dataloader` :)
If you have set `shuffle=False`, then you will just repeat the same 2 batches. And if you have set  `shuffle=False` it depends on the implementation details which I don't know. If `val_dataloader` created before each validation loop, then first 2 batches and second 2 batches will be totally independent (it may be other 2 batches, or it may be the same two batches, or they can have only one common batch). And if `val_dataloader` is replaced only when exhausted, then you will take 4 of 10.

**Comment by kuynzereb** [comment]
Well, this PR fixes computation of the total number of batches during one training epoch. Total number of batches is equal to the sum of all training batches and all validation batches. All validation batches is the product of the number of validation checks per this epoch and the number of validation batches during one validation loop.

The number of validation checks per epoch is determined by the parameter `val_check_interval`. By setting this you control how often you want to check validation metrics. The number of validation batches during one validation loop is determined by the parameter `val_percent_check`. By setting this you control how accurate and how fast your estimations will be. The smaller `val_percent_check` the more accurate and the slower estimates will be.

And if you have very small training epochs you can also set `check_val_every_n_epoch` to check validation only every `check_val_every_n_epoch`-th epoch.

**Comment by Borda** [comment]
> The number of validation checks per epoch is determined by the parameter `val_check_interval`. By setting this you control how often you want to check validation metrics. The number of validation batches during one validation loop is determined by the parameter `val_percent_check`. By setting this you control how accurate and how fast your estimations will be. The smaller `val_percent_check` the more accurate and the slower estimates will be.
> 
> And if you have very small training epochs you can also set `check_val_every_n_epoch` to check validation only every `check_val_every_n_epoch`-th epoch.

nice, pls add this to the code as a comment, so if someone in future will toggle this, he will know too...
@williamFalcon it looks fine to me :)

**Comment by kuynzereb** [comment]
Hm, actually @williamFalcon has already added some brief comments to the code. As for my last comment it seems for me that it is more appropriate for the docs but not for the code (because it is more about usage but not about implementation). And maybe the code will be more self-explaining if we rename `val_check_batch` to `check_val_every_n_batch` (by analogy with `check_val_every_n_epoch`) . What do you think?
